### PR TITLE
e2e, storage-migration: Replace client Update operations with patch

### PR DESCRIPTION
**Pull request by ormergi**
### What this PR does
Before this PR:
 E2e storage migration helpers `UpdateVMWithPVC` and `UpdateVMWithDV` use client Update operation   to mutate the test VM object.

After this PR:
 E2e storage migration helpers `UpdateVMWithPVC` and `UpdateVMWithDV` use client Patch to mutate the test VM object.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Special notes for your reviewer
The code that finds the index of the subject volume name is simplified by using builtin `slices`.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by ormergi** [comment]
Rebased

File: tests/storage/migration.go
Comment (ormergi) [reply tests/storage/migration.go]: > Shouldn't this be /spec/template/spec/volumes/%d/volumeSource/persistentVolumeClaim?

Yes, this is the right format, thanks.

> I would personally replace the entire volume without the need of the parsing of the claimname

From what I understand the PVC name is given (claim name) by the test setup, and the subject volume source is updated with new the given claim name.
The claim name is not being parsed, but it does converted to fit volume source JSON format.

How would it work if volume is replaced? could you please elaborate or give example?

**Review by ormergi — COMMENTED** [review]
[Changes](https://github.com/kubevirt/kubevirt/compare/d08298edb23a9d260af0c19a5622e1845c25989c..2a6615ee62e30050f3633523b13d69f88d365cfc): resolve https://github.com/kubevirt/kubevirt/pull/12478/commits/35a8a14fab509fb11c70af05979492a1f8ccb587#r1697035752

@alicefr thanks for the feedback, please have another look :pray:

File: tests/storage/migration.go
Comment (alicefr) [reply tests/storage/migration.go]: Well, the claim is the only value set in the volume, so you basically replace the entire struct. But the 2 are achieving the same

Comment (alicefr) [reply tests/storage/migration.go]: Sorry, that's wrong. If you complete patch the volume, then you also set the dv to nil

File: tests/storage/migration.go
Comment (alicefr) [inline tests/storage/migration.go:236]: nit: can you please add the same print as the function above

**Comment by alicefr** [comment]
@ormergi  in this case, I don't think it makes a difference, but I would add a test operation before the replace, in case the field is changed by something else

**Comment by alicefr** [comment]
2 nits, otherwise the changes look great
/approve
/hold
to adjust the comments

**Comment by alicefr** [comment]
/unhold
Looks good

**Review by akalenyu — APPROVED** [review]

**Comment by akalenyu** [comment]
/lgtm cancel
```
07:51:58: 07:51:58: tests/storage/migration.go:413:19: cannot use vm.Name (variable of type string) as *"kubevirt.io/api/core/v1".VirtualMachine value in argument to updateVMWithDV
```

**Comment by ormergi** [comment]
Rebased

File: tests/storage/migration.go
Comment (RamLavi) [inline tests/storage/migration.go:380]: This should be in the former commit

**Review by RamLavi — COMMENTED** [review]
small change, otherwise lgtm

File: tests/storage/migration.go
Comment (ormergi) [reply tests/storage/migration.go:380]: Thanks, fixed.

**Comment by ormergi** [comment]
[Changes](https://github.com/kubevirt/kubevirt/compare/45a18aa2ce17c7414626a1895d3fef629dae54ff..5e80317205408e66ec2f45b03582dcf850fc7360): changes since last review cycle (up to this comment https://github.com/kubevirt/kubevirt/pull/12478#issuecomment-2316501841): 
- Fixed volume index assersion
- Removed lefover volume index var, it was mistakenly used. 
- Fix call to changes functions, passing VM object instead of name.
- Use Volume object for replacing between DV and PVC instead of patching the VolumeSource directly following this thread https://github.com/kubevirt/kubevirt/pull/12478#discussion_r1697042849

**Comment by ormergi** [comment]
/hold

Placing hold as there has been changes since last review from the approver.
Should be removed once changes are acknowledged by the approver.

**Comment by ormergi** [comment]
@akalenyu could you please have another look?

**Comment by ormergi** [comment]
[Changes](https://github.com/kubevirt/kubevirt/compare/45a18aa2ce17c7414626a1895d3fef629dae54ff..98f52f6c6edacb35687a5b6b3dac6ae2e3b4a702): 
changes since last review cycle (up to this comment https://github.com/kubevirt/kubevirt/pull/12478#issuecomment-2316501841):
-    Fixed volume index assersion
-    Removed leftover volume index var, it was mistakenly used.
-    Fix call to changes functions, passing VM object instead of name.
-    Use Volume object for replacing between DV and PVC instead of patching the VolumeSource directly following this thread https://github.com/kubevirt/kubevirt/pull/12478#discussion_r1697042849
- fixed dateVolumeTemplate patch path (missing metadata)

**Comment by alicefr** [comment]
Looks good
/lgtm
/unhold

**Pull request merged**
